### ISSUE TITLE
Add an option to skip SSL verification for users behind firewalls

### ISSUE THREAD

**Issue by paul-gauthier** [Aider-AI/aider#664]
### Issue

Per twitter thread:

https://x.com/Liu_eroteme/status/1800186374039187525

**Comment by dannylank** [comment]
I change in ..\Lib\site-packages\httpx\_transports\default.py change lines  ssl_context = create_ssl_context(verify=False: Verify to False. ( lines 136 and 277)

**Comment by paul-gauthier** [comment]
I just added a `--no-verify-ssl` option, which disables it per the litellm docs:

https://docs.litellm.ai/docs/providers/openai#set-ssl_verifyfalse

**Comment by paul-gauthier** [comment]
I'm going to close this issue for now, but feel free to add a comment here and I will re-open or file a new issue any time.

**Comment by dannylank** [comment]
Installing collected packages: aider-chat
  Attempting uninstall: aider-chat
    Found existing installation: aider-chat 0.38.0
    Uninstalling aider-chat-0.38.0:
      Successfully uninstalled aider-chat-0.38.0
Successfully installed aider-chat-0.39.0 <-----------------<<

aider: error: unrecognized arguments: --no-verify-ssl

**Comment by paul-gauthier** [comment]
The change is available in the main branch. You can get it by installing the latest version from github:

```
python -m pip install --upgrade git+https://github.com/paul-gauthier/aider.git
```

If you have a chance to try it, let me know if it works better for you.

**Comment by dannylank** [comment]
aider --git --browser --no-auto-commits --opus --no-verify-ssl                         

CONTROL-C to exit...
Watching for file changes.

  You can now view your Streamlit app in your browser.

  Local URL: http://localhost:8501
  Network URL: http://192.168.1.7:8501

Error checking pypi for new version: HTTPSConnectionPool(host='pypi.org', port=443): Max retries exceeded with url:
/pypi/aider-chat/json (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate
verify failed: self-signed certificate in certificate chain (_ssl.c:1000)')))
**Aider v0.40.2-dev**
Models: claude-3-opus-20240229 with diff edit format, weak model claude-3-haiku-20240307
Git repo: .git with 2 files
Repo-map: using 1024 tokens
litellm.APIConnectionError: HTTPSConnectionPool(host='api.anthropic.com', port=443): Max retries exceeded with url: /v1/messages (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate in certificate chain (_ssl.c:1000)')))

**Comment by Neurrone** [comment]
get the same error as well even though I've set this option to false.

**Comment by paul-gauthier** [comment]
Thanks for reporting this regression. It should be fixed in the main branch.

The change is available in the main branch. You can get it by installing the latest version from github:

```
python -m pip install --upgrade git+https://github.com/paul-gauthier/aider.git
```

If you have a chance to try it, let me know if it works better for you.

**Comment by Liu-Eroteme** [comment]
Aiders update check does respect this option, but the anthropic API does not:

```
> aider --no-verify-ssl

Aider v0.54.7
Main model: claude-3-5-sonnet-20240620 with diff edit format, infinite output
Weak model: claude-3-haiku-20240307
Git repo: .git with 7 files
Repo-map: using 1024 tokens, auto refresh
Use /help <question> for help, run "aider --help" to see cmd line args
───────────────────────────────────────────────────────────────────────────────────────────────────────────────
> test!

litellm.InternalServerError: AnthropicException - [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed:
self-signed certificate in certificate chain (_ssl.c:1007). Handle with `litellm.InternalServerError`.
Retrying in 0.2 seconds...
litellm.InternalServerError: AnthropicException - [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed:
self-signed certificate in certificate chain (_ssl.c:1007). Handle with `litellm.InternalServerError`.
Retrying in 0.5 seconds...
litellm.InternalServerError: AnthropicException - [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed:
self-signed certificate in certificate chain (_ssl.c:1007). Handle with `litellm.InternalServerError`.
Retrying in 1.0 seconds...



^C again to exit
───────────────────────────────────────────────────────────────────────────────────────────────────────────────
>
```

**Comment by QiBaobin** [comment]
Got below error too while using gemini:

litellm.APIConnectionError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)

**Comment by paul-gauthier** [comment]
What version of aider?

**Comment by QiBaobin** [comment]
0.56, and making a change to site-packages\httpx_transports\default.py work, but need some retries.

**Comment by paul-gauthier** [comment]
The fix is available in the main branch. You can get it by installing the latest version from github:

```
aider --install-main-branch

# or...

python -m pip install --upgrade git+https://github.com/paul-gauthier/aider.git
```

If you have a chance to try it, let me know if it works better for you.

**Comment by paul-gauthier** [comment]
/
I'm going to close this issue for now, but feel free to add a comment here and I will re-open. Or feel free to file a new issue any time.

**Comment by QiBaobin** [comment]
thanks @paul-gauthier , just verified on 0.59.1, it works now.

**Comment by spaasis** [comment]
I'm still hitting this with 0.59.1:

```
aider --model ollama/llama3.1:latest --no-verify-ssl
──────────────────────────────────
Aider v0.59.1
Model: ollama/llama3.1:latest with whole edit format
──────────────────────────────────
> Hello

[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1000)
Retrying in 0.2 seconds...

```

I'm attempting to connect to a company internal OpenWebUI instance with a self signed certificate, using
`OLLAMA_API_BASE` = https://myaiservice.domain.com/ollama

**Comment by paul-gauthier** [comment]
This seems to be an issue with litellm's support for ollama.

**Comment by paul-gauthier** [comment]
This is a litellm bug for Ollama models.

https://github.com/BerriAI/litellm/issues/6499

We need them to address it or provide guidance.

**Comment by paul-gauthier** [comment]
This is waiting on litellm lookin at my PR: https://github.com/BerriAI/litellm/pull/7028

**Comment by paul-gauthier** [comment]
Sounds like this will be fixed in litellm soon.

**Comment by paul-gauthier** [comment]
Using `aider --no-verify-ssl` seems to work now for everything except `ollama_chat/` models. Even `ollama/` models work.

I wonder if the problem with `ollama_chat/` is related to: https://github.com/BerriAI/litellm/issues/8333

**Comment by daveLeiser** [comment]
It looks like the latest version of Aider (0.74.2) no longer respects the --no-verify-ssl option.

I've been using Aider for a few months and recently set up a new MacBook. Since I work within a company network, I need to disable SSL verification. I enabled verify-ssl: false in .aider.conf.yaml and also tried launching Aider with:

`aider --cache-prompts --no-stream --no-verify-ssl --sonnet --anthropic-api-key ***`

I'm using the Claude 3.5 Sonnet model. On my new MacBook, Aider fails with an SSL certificate verification error. To verify, I tested the same setup on my old MacBook, where Aider v0.73.x worked fine. However, after upgrading to v0.74.2, it stopped working there as well.

Aider starts successfully, but when sending a prompt, I get this error:

`litellm.InternalServerError: AnthropicException - [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1000). Handle with 'litellm.InternalServerError'.`

Is this a known issue in v0.74.2, or is there a configuration change I may have missed?

**Comment by jufo** [comment]
This is still happening in 0.75.1:

```powershell
> aider --no-verify-ssl --model sonnet
────────────────────────────────────────────────────────────────────────────────────────────────────────────
HTTPSConnectionPool(host='raw.githubusercontent.com', port=443): Max retries exceeded with url: /BerriAI/litellm/main/model_prices_and_context_window.json (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1000)')))
Aider v0.75.1
...
```

**Comment by jufo** [comment]
In my case, where the problem is caused by corporate use of ZScaler, I found (at last) that this is all I need:

`> python -m pip install pip-system-certs`

Now in VS Code's Terminal (Powershell, Windows 11):

```powershell
> aider --model sonnet
─────────────────────────────────────────────────────────────────────────────────────────
Aider v0.75.1
Main model: anthropic/claude-3-7-sonnet-20250219 with diff edit format, infinite output
Weak model: anthropic/claude-3-5-haiku-20241022
Git repo: .git with 13 files
Repo-map: using 4096 tokens, auto refresh
─────────────────────────────────────────────────────────────────────────────────────────
>
```

**Comment by spaasis** [comment]
I worked around this by creating an nginx proxy:

```
events {}

http {
    server {
        listen 80;
        server_name ollama.local;

        location / {
            proxy_pass https://org-internal-ollama;
            proxy_set_header Authorization "Bearer eyJh....";
            proxy_ssl_verify off;  # Disable SSL verification for self-signed certificates
        }
    }
}

```

**Comment by themastersheep** [comment]
Whilst not entirely the same issue, I am not only sat behind a firewall but we use internal mirrors for "approved" software installs.  As such we have our own internal "pypi" mirror updated independently.  

The `check_version` function directly calls the public mirror (https://github.com/Aider-AI/aider/blob/main/aider/versioncheck.py#L78)

To install I updated the `PIP_INDEX_URL` and `UV_DEFAULT_INDEX` environment variables, could the `check_version` be updated to be configured in a similar way ?

**Comment by Mattwmaster58** [comment]
Adding my experience - `--no-verify-ssl` didn't do anything for me. Installing in the same venv (`pipx inject aider-chat pip-system-certs`) did work.

**Comment by azinsharaf** [comment]
> Adding my experience - `--no-verify-ssl` didn't do anything for me. Installing in the same venv (`pipx inject aider-chat pip-system-certs`) did work.

this resolved the ssl cert issue. do you know what it does?

**Comment by amp9020** [comment]
I'm behind a corp firewall. 
After spending the majority of the day trying to get past this, none of these solutions worked.

After some deep diving, here's what worked for me:

Installing or upgrading pip-system-certs using:

**python -m pip install pip-system-certs --use-feature=truststore**

I already tried installing pip-system-certs but was not aware of the --use-feature param.

The difference was the --use=feature param.

After I ran that, can run aider no problem.

I also removed the --no-verify-ssl param and verified this works without this param.

Cheers!

**Comment by holysheep123** [comment]
For developers in China where direct access to OpenAI/Anthropic/Google APIs is blocked, the SSL workarounds above won't help — the issue is network-level blocking, not SSL certificates.

What works: use an API relay that proxies requests through overseas servers.

**Option 1: One-command setup**
```bash
npx @simonyea/holysheep-cli setup
```
This auto-configures Aider (and Claude Code, Codex, Gemini CLI, Cursor) with the correct `OPENAI_BASE_URL` / `ANTHROPIC_BASE_URL`.

**Option 2: Manual**
```bash
export OPENAI_API_KEY=your_key
export OPENAI_BASE_URL=https://api.holysheep.ai/v1
# or for Anthropic models:
export ANTHROPIC_API_KEY=your_key
export ANTHROPIC_BASE_URL=https://api.holysheep.ai
```

The relay at [holysheep.ai](https://holysheep.ai) supports all models (GPT-5, Claude, Gemini) with pay-as-you-go pricing. This is specifically useful for the China firewall case that several people here seem to be running into.
